[AMDGPU] add tests for loop definition of bitconvert

Shoreshen opened the pull request:
This PR add test cases for all types of bit conversation, it prepares for PR: https://github.com/llvm/llvm-project/pull/132899



All tests passed due to:
1. For DAG, pattern will not separate SReg and VReg. One of the sample is:
    ```
   define <2 x double> @v_bitcast_v4f32_to_v2f64(<4 x float> inreg %a, i32 %b) {
     %cmp = icmp eq i32 %b, 0
     br i1 %cmp, label %cmp.true, label %cmp.false
   
   cmp.true:
     %a1 = fadd <4 x float> %a, splat (float 1.000000e+00)
     %a2 = bitcast <4 x float> %a1 to <2 x double>
     br label %end
   
   cmp.false:
     %a3 = bitcast <4 x float> %a to <2 x double>
     br label %end
   
   end:
     %phi = phi <2 x double> [ %a2, %cmp.true ], [ %a3, %cmp.false ]
     ret <2 x double> %phi
   }
   ```
   It suppose to select from scalar register patterns. But the Vreg pattern is matched is as follow:
    ```
   Debug log:
   ISEL: Starting selection on root node: t3: v2f64 = bitcast t2
   ISEL: Starting pattern match
     Initial Opcode index to 440336
     Skipped scope entry (due to false predicate) at index 440339, continuing at 440367
     Skipped scope entry (due to false predicate) at index 440368, continuing at 440396
     Skipped scope entry (due to false predicate) at index 440397, continuing at 440435
     Skipped scope entry (due to false predicate) at index 440436, continuing at 440467
     Skipped scope entry (due to false predicate) at index 440468, continuing at 440499
     Skipped scope entry (due to false predicate) at index 440500, continuing at 440552
     Skipped scope entry (due to false predicate) at index 440553, continuing at 440587
     Skipped scope entry (due to false predicate) at index 440588, continuing at 440622
     Skipped scope entry (due to false predicate) at index 440623, continuing at 440657
     Skipped scope entry (due to false predicate) at index 440658, continuing at 440692
     Skipped scope entry (due to false predicate) at index 440693, continuing at 440727
     Skipped scope entry (due to false predicate) at index 440728, continuing at 440769
     Skipped scope entry (due to false predicate) at index 440770, continuing at 440798
     Skipped scope entry (due to false predicate) at index 440799, continuing at 440836
     Skipped scope entry (due to false predicate) at index 440837, continuing at 440870
     TypeSwitch[v2f64] from 440873 to 440892
   
   Patterns:
   /*440892*/    OPC_CompleteMatch, 1, 0, 
                  // Src: (bitconvert:{ *:[v2f64] } VReg_128:{ *:[v4f32] }:$src0) - Complexity = 3
                  // Dst: VReg_128:{ *:[v2f64] }:$src0
    ```
2. Global isel will use `Select_COPY` to select bitcast

Shoreshen committed:
add tests for loop definition of bitconvert

Shoreshen committed:
fix lit

Pierre-vh commented:
Please add `[AMDGPU]` prefix to the title

Also what are these lit tests for ? why do we need to test all widths?

Shoreshen commented:
> Please add `[AMDGPU]` prefix to the title
> 
> Also what are these lit tests for ? why do we need to test all widths?

Hi @Pierre-vh , this is adding test PR for https://github.com/llvm/llvm-project/pull/132899, which using loop in tablegen to generate bitconvert patterns, instead of doing it one by one.... Thanks~

Shoreshen committed:
Merge branch 'main' into add-loop-bitconvert-tests

arsenm commented:
> Also what are these lit tests for ? why do we need to test all widths?

Every operation with every type should be tested.

For this case, we already should have test coverage so I'm not sure I understand what is new here. We should perhaps split / replace the existing test with a more consistent pattern, but this patch is purely additive?

arsenm commented on llvm/test/CodeGen/AMDGPU/amdgcn.bitcast.160bit.ll:
```diff
@@ -0,0 +1,178 @@
+; RUN: llc -mtriple=amdgcn -mcpu=gfx900 -amdgpu-codegenprepare-break-large-phis-threshold=4096 < %s | FileCheck -check-prefixes=GFX9 %s
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1100 -amdgpu-codegenprepare-break-large-phis-threshold=4096 < %s | FileCheck -check-prefixes=GFX11 %s
+
+define <5 x float> @v_bitcast_v5i32_to_v5f32(<5 x i32> %a, i32 %b) {
```
Doesn't have scalar version

arsenm commented on llvm/test/CodeGen/AMDGPU/amdgcn.bitcast.160bit.ll:
```diff
@@ -0,0 +1,178 @@
+; RUN: llc -mtriple=amdgcn -amdgpu-codegenprepare-break-large-phis-threshold=4096 < %s | FileCheck -check-prefix=GCN %s
+; RUN: llc -mtriple=amdgcn -mcpu=tonga -amdgpu-codegenprepare-break-large-phis-threshold=4096 < %s | FileCheck -check-prefixes=VI %s
+; RUN: llc -mtriple=amdgcn -mcpu=gfx900 -amdgpu-codegenprepare-break-large-phis-threshold=4096 < %s | FileCheck -check-prefixes=GFX9 %s
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1100 -amdgpu-codegenprepare-break-large-phis-threshold=4096 < %s | FileCheck -check-prefixes=GFX11 %s
```
can you avoid adding special flags

Shoreshen replied:
Hi @arsenm , by avoid special flags you mean change from `FileCheck -check-prefix=GCN %s` to `FileCheck %s` ??

shiltian replied:
I think @arsenm meant those like `-amdgpu-codegenprepare-break-large-phis-threshold=4096`.

Shoreshen replied:
Hi thanks @shiltian , flags have been removed

Shoreshen committed:
remove old test

Shoreshen committed:
Merge branch 'main' into add-loop-bitconvert-tests

Shoreshen commented:
> > Also what are these lit tests for ? why do we need to test all widths?
> 
> Every operation with every type should be tested.
> 
> For this case, we already should have test coverage so I'm not sure I understand what is new here. We should perhaps split / replace the existing test with a more consistent pattern, but this patch is purely additive?

Hi @arsenm , old test has been removed~~

Shoreshen committed:
replace bitconvert tests

Shoreshen committed:
Merge branch 'main' into add-loop-bitconvert-tests

Shoreshen commented:
> > Also what are these lit tests for ? why do we need to test all widths?
> 
> Every operation with every type should be tested.
> 
> For this case, we already should have test coverage so I'm not sure I understand what is new here. We should perhaps split / replace the existing test with a more consistent pattern, but this patch is purely additive?

Hi @arsenm , I removed the old tests and split them into different files named by bit width.

Shoreshen committed:
Merge remote-tracking branch 'origin/main' into add-loop-bitconvert-tests

Shoreshen committed:
remove options

Shoreshen committed:
Merge branch 'main' into add-loop-bitconvert-tests

Shoreshen commented:
Hi @shiltian @arsenm ,  just ask if this PR is OK. Any fix up need, please comment~ Thanks~

Shoreshen committed:
Merge branch 'main' into add-loop-bitconvert-tests

Shoreshen committed:
Merge branch 'main' into add-loop-bitconvert-tests

Shoreshen committed:
Merge branch 'main' into add-loop-bitconvert-tests

Shoreshen committed:
Merge branch 'main' into add-loop-bitconvert-tests

Shoreshen committed:
Merge branch 'main' into add-loop-bitconvert-tests

arsenm commented on llvm/test/CodeGen/AMDGPU/amdgcn.bitcast.ptr.ll:
```diff
@@ -0,0 +1,68 @@
+; RUN: llc -mtriple=amdgcn -mcpu=gfx900  < %s | FileCheck -check-prefixes=GFX9 %s
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1100  < %s | FileCheck -check-prefixes=GFX11 %s
+
+define amdgpu_kernel void @bitcast_i8ptr_v16i8ptr(ptr addrspace(1) %out, ptr addrspace(1) %in) {
```
Why "ptr" in the name? There is no pointer cast here. This is also not testing the reverse direction, or all of the interesting type sizes. I think we should be generating all of the cases

Shoreshen replied:
Hi @arsenm , this case is a copy from the old file llvm-`project/llvm/test/CodeGen/AMDGPU/amdgcn.bitcast.ll: L56~L116`:
```
define amdgpu_kernel void @i8ptr_v16i8ptr(ptr addrspace(1) %out, ptr addrspace(1) %in) {
; GCN-LABEL: i8ptr_v16i8ptr:
; GCN:       ; %bb.0: ; %entry
; GCN-NEXT:    s_load_dwordx4 s[0:3], s[4:5], 0x9
; GCN-NEXT:    s_waitcnt lgkmcnt(0)
; GCN-NEXT:    s_load_dwordx4 s[4:7], s[2:3], 0x0
; GCN-NEXT:    s_mov_b32 s3, 0xf000
; GCN-NEXT:    s_mov_b32 s2, -1
; GCN-NEXT:    s_waitcnt lgkmcnt(0)
; GCN-NEXT:    v_mov_b32_e32 v0, s4
; GCN-NEXT:    v_mov_b32_e32 v1, s5
; GCN-NEXT:    v_mov_b32_e32 v2, s6
; GCN-NEXT:    v_mov_b32_e32 v3, s7
; GCN-NEXT:    buffer_store_dwordx4 v[0:3], off, s[0:3], 0
; GCN-NEXT:    s_endpgm
;
; VI-LABEL: i8ptr_v16i8ptr:
; VI:       ; %bb.0: ; %entry
; VI-NEXT:    s_load_dwordx4 s[0:3], s[4:5], 0x24
; VI-NEXT:    s_waitcnt lgkmcnt(0)
; VI-NEXT:    s_load_dwordx4 s[4:7], s[2:3], 0x0
; VI-NEXT:    v_mov_b32_e32 v4, s0
; VI-NEXT:    v_mov_b32_e32 v5, s1
; VI-NEXT:    s_waitcnt lgkmcnt(0)
; VI-NEXT:    v_mov_b32_e32 v0, s4
; VI-NEXT:    v_mov_b32_e32 v1, s5
; VI-NEXT:    v_mov_b32_e32 v2, s6
; VI-NEXT:    v_mov_b32_e32 v3, s7
; VI-NEXT:    flat_store_dwordx4 v[4:5], v[0:3]
; VI-NEXT:    s_endpgm
;
; GFX9-LABEL: i8ptr_v16i8ptr:
; GFX9:       ; %bb.0: ; %entry
; GFX9-NEXT:    s_load_dwordx4 s[0:3], s[4:5], 0x24
; GFX9-NEXT:    v_mov_b32_e32 v4, 0
; GFX9-NEXT:    s_waitcnt lgkmcnt(0)
; GFX9-NEXT:    s_load_dwordx4 s[4:7], s[2:3], 0x0
; GFX9-NEXT:    s_waitcnt lgkmcnt(0)
; GFX9-NEXT:    v_mov_b32_e32 v0, s4
; GFX9-NEXT:    v_mov_b32_e32 v1, s5
; GFX9-NEXT:    v_mov_b32_e32 v2, s6
; GFX9-NEXT:    v_mov_b32_e32 v3, s7
; GFX9-NEXT:    global_store_dwordx4 v4, v[0:3], s[0:1]
; GFX9-NEXT:    s_endpgm
;
; GFX11-LABEL: i8ptr_v16i8ptr:
; GFX11:       ; %bb.0: ; %entry
; GFX11-NEXT:    s_load_b128 s[0:3], s[4:5], 0x24
; GFX11-NEXT:    s_waitcnt lgkmcnt(0)
; GFX11-NEXT:    s_load_b128 s[4:7], s[2:3], 0x0
; GFX11-NEXT:    v_mov_b32_e32 v4, 0
; GFX11-NEXT:    s_waitcnt lgkmcnt(0)
; GFX11-NEXT:    v_dual_mov_b32 v0, s4 :: v_dual_mov_b32 v3, s7
; GFX11-NEXT:    v_dual_mov_b32 v1, s5 :: v_dual_mov_b32 v2, s6
; GFX11-NEXT:    global_store_b128 v4, v[0:3], s[0:1]
; GFX11-NEXT:    s_endpgm
entry:
  %0 = load <16 x i8>, ptr addrspace(1) %in
  store <16 x i8> %0, ptr addrspace(1) %out
  ret void
}
```

I'm kind of not sure how I should handle this because there is no bitcast.....

arsenm commented on llvm/test/CodeGen/AMDGPU/amdgcn.bitcast.ptr.ll:
```diff
@@ -0,0 +1,68 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+
+; RUN: llc -mtriple=amdgcn  < %s | FileCheck -check-prefix=GCN %s
```
Avoid using the default target. -mcpu=tahiti? Also extra spaces before <

Shoreshen replied:
Hi @arsenm , this case is a copy from the old file llvm-`project/llvm/test/CodeGen/AMDGPU/amdgcn.bitcast.ll: L56~L116`:
```
define amdgpu_kernel void @i8ptr_v16i8ptr(ptr addrspace(1) %out, ptr addrspace(1) %in) {
; GCN-LABEL: i8ptr_v16i8ptr:
; GCN:       ; %bb.0: ; %entry
; GCN-NEXT:    s_load_dwordx4 s[0:3], s[4:5], 0x9
; GCN-NEXT:    s_waitcnt lgkmcnt(0)
; GCN-NEXT:    s_load_dwordx4 s[4:7], s[2:3], 0x0
; GCN-NEXT:    s_mov_b32 s3, 0xf000
; GCN-NEXT:    s_mov_b32 s2, -1
; GCN-NEXT:    s_waitcnt lgkmcnt(0)
; GCN-NEXT:    v_mov_b32_e32 v0, s4
; GCN-NEXT:    v_mov_b32_e32 v1, s5
; GCN-NEXT:    v_mov_b32_e32 v2, s6
; GCN-NEXT:    v_mov_b32_e32 v3, s7
; GCN-NEXT:    buffer_store_dwordx4 v[0:3], off, s[0:3], 0
; GCN-NEXT:    s_endpgm
;
; VI-LABEL: i8ptr_v16i8ptr:
; VI:       ; %bb.0: ; %entry
; VI-NEXT:    s_load_dwordx4 s[0:3], s[4:5], 0x24
; VI-NEXT:    s_waitcnt lgkmcnt(0)
; VI-NEXT:    s_load_dwordx4 s[4:7], s[2:3], 0x0
; VI-NEXT:    v_mov_b32_e32 v4, s0
; VI-NEXT:    v_mov_b32_e32 v5, s1
; VI-NEXT:    s_waitcnt lgkmcnt(0)
; VI-NEXT:    v_mov_b32_e32 v0, s4
; VI-NEXT:    v_mov_b32_e32 v1, s5
; VI-NEXT:    v_mov_b32_e32 v2, s6
; VI-NEXT:    v_mov_b32_e32 v3, s7
; VI-NEXT:    flat_store_dwordx4 v[4:5], v[0:3]
; VI-NEXT:    s_endpgm
;
; GFX9-LABEL: i8ptr_v16i8ptr:
; GFX9:       ; %bb.0: ; %entry
; GFX9-NEXT:    s_load_dwordx4 s[0:3], s[4:5], 0x24
; GFX9-NEXT:    v_mov_b32_e32 v4, 0
; GFX9-NEXT:    s_waitcnt lgkmcnt(0)
; GFX9-NEXT:    s_load_dwordx4 s[4:7], s[2:3], 0x0
; GFX9-NEXT:    s_waitcnt lgkmcnt(0)
; GFX9-NEXT:    v_mov_b32_e32 v0, s4
; GFX9-NEXT:    v_mov_b32_e32 v1, s5
; GFX9-NEXT:    v_mov_b32_e32 v2, s6
; GFX9-NEXT:    v_mov_b32_e32 v3, s7
; GFX9-NEXT:    global_store_dwordx4 v4, v[0:3], s[0:1]
; GFX9-NEXT:    s_endpgm
;
; GFX11-LABEL: i8ptr_v16i8ptr:
; GFX11:       ; %bb.0: ; %entry
; GFX11-NEXT:    s_load_b128 s[0:3], s[4:5], 0x24
; GFX11-NEXT:    s_waitcnt lgkmcnt(0)
; GFX11-NEXT:    s_load_b128 s[4:7], s[2:3], 0x0
; GFX11-NEXT:    v_mov_b32_e32 v4, 0
; GFX11-NEXT:    s_waitcnt lgkmcnt(0)
; GFX11-NEXT:    v_dual_mov_b32 v0, s4 :: v_dual_mov_b32 v3, s7
; GFX11-NEXT:    v_dual_mov_b32 v1, s5 :: v_dual_mov_b32 v2, s6
; GFX11-NEXT:    global_store_b128 v4, v[0:3], s[0:1]
; GFX11-NEXT:    s_endpgm
entry:
  %0 = load <16 x i8>, ptr addrspace(1) %in
  store <16 x i8> %0, ptr addrspace(1) %out
  ret void
}
```

I'm kind of not sure how I should handle this because there is no bitcast.....

arsenm commented on llvm/test/CodeGen/AMDGPU/amdgcn.bitcast.ll:
Currently not convinced we aren't losing test coverage. The content of some of the new bitwidth suffixed files don't have enough functions to justify splitting up. I'd prefer generated, wholly consistent tests with the suffixes if we're splitting this up
